crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -291,6 +322,94 @@ impl Visitor<'_> for ReboundComprehensionVisitor<'_> {
+    fn emit_errors(mut self) {
+        self.names.sort_by_key(|name| &name.id);
+        for (n1, n2) in self.names.iter().zip(self.names.iter().skip(1)) {
+            if n1.id == n2.id {
+                SemanticSyntaxChecker::add_error(
+                    self.ctx,
+                    SemanticSyntaxErrorKind::MultipleCaseAssignment(n2.id.clone()),
+                    n2.range,
+                );
+            }
+        }
+    }
+
+    fn visit_pattern(&mut self, pattern: &'a Pattern) {
+        // test_err multiple_assignment_in_case_pattern
+        // match 2:
+        //     case x as x: ...  # MatchAs
+        //     case [y, z, y]: ...  # MatchSequence
+        //     case [y, z, *y]: ...  # MatchSequence
+        //     case [y, y, y]: ...  # MatchSequence multiple
+        //     case {1: x, 2: x}: ...  # MatchMapping duplicate pattern
+        //     case {1: x, **x}: ...  # MatchMapping duplicate in **rest
+        //     case Class(x, x): ...  # MatchClass positional
+        //     case Class(x=1, x=2): ...  # MatchClass keyword
+        //     case [x] | {1: x} | Class(x=1, x=2): ...  # MatchOr
+
+        // test_ok multiple_assignment_in_case_pattern
+        // match 2:
+        //     case Class(x) | [x] | x: ...
+        match pattern {
+            Pattern::MatchValue(_) | Pattern::MatchSingleton(_) => {}
+            Pattern::MatchStar(ast::PatternMatchStar { name, .. }) => {
+                if let Some(name) = name {
+                    self.names.push(name);
+                }
+            }
+            Pattern::MatchSequence(ast::PatternMatchSequence { patterns, .. }) => {
+                for pattern in patterns {
+                    self.visit_pattern(pattern);
+                }
+            }
+            Pattern::MatchMapping(ast::PatternMatchMapping { patterns, rest, .. }) => {
+                for pattern in patterns {
+                    self.visit_pattern(pattern);
+                }
+                if let Some(rest) = rest {
+                    self.names.push(rest);
+                }
+            }
+            Pattern::MatchClass(ast::PatternMatchClass { arguments, .. }) => {
+                for pattern in &arguments.patterns {
+                    self.visit_pattern(pattern);
+                }
+                for keyword in &arguments.keywords {
+                    self.names.push(&keyword.attr);
+                    self.visit_pattern(&keyword.pattern);
+                }
+            }
+            Pattern::MatchAs(ast::PatternMatchAs { pattern, name, .. }) => {
+                if let Some(pattern) = pattern {
+                    self.visit_pattern(pattern);
+                }
+                if let Some(name) = name {
+                    self.names.push(name);
+                }
+            }
+            Pattern::MatchOr(ast::PatternMatchOr { patterns, .. }) => {
+                // each of these patterns should be visited separately
```

Comment:
I think it would be useful to provide the reason for why should this be done. From what I understand, I think it's because they're separate patterns and checking for duplicate names are isolated to a single pattern.